### PR TITLE
Bottom-up pattern matching

### PR DESCRIPTION
**Pull request by Kha**
Patterns for now: inc, dec, address modes

-20% on fannkuch \o/

### CODE REVIEW

File: src/main/scala/mjis/CodeGenerator.scala [hunk 1]
```diff
@@ -49,27 +49,48 @@ class CodeGenerator(a: Unit) extends Phase[AsmProgram] {
...
 
       basicBlocks.values.foreach(generateInstructionsForPhi)
 
-      function.epilogue.controlFlowInstructions += Ret()
+      // spill param registers early and in a fixed order
+      for (reg <- ParamRegisters)
+        usedParamRegisters.get(reg) match {
+          case Some(n) => function.prologue.instructions += Mov(RegisterOperand(reg, n.getMode.getSizeBytes), regOp(n))
+          case None =>
+        }
```
Comment (TehMillhouse) [inline src/main/scala/mjis/CodeGenerator.scala]: {nit} (reminder that you could use a `map` here instead of a single-case-pattern-match.)

Comment (Kha) [inline src/main/scala/mjis/CodeGenerator.scala]: Thanks, replaced with `Option.foreach`

**Comment by TehMillhouse** [comment]
Légume to me!

**Comment by Stoeoef** [comment]
+1

**Pull request merged**